diff --git a/llama.cpp b/llama.cpp
index 5329bd8..fa2b3d2 100644
--- a/llama.cpp
+++ b/llama.cpp
@@ -9625,7 +9625,9 @@ static void llama_log_internal(ggml_log_level level, const char * format, ...) {
 
 static void llama_log_callback_default(ggml_log_level level, const char * text, void * user_data) {
     (void) level;
-    (void) user_data;
-    fputs(text, stderr);
-    fflush(stderr);
+    bool * enable_log = static_cast<bool *>(user_data);
+    if (enable_log && *enable_log) {
+        fputs(text, stderr);
+        fflush(stderr);
+    }
 }
