--- a/gcc/splay-tree-utils.h	2021-05-14 06:09:43.289274290 +0000
+++ b/gcc/splay-tree-utils.h	2021-05-14 06:24:13.628159368 +0000
@@ -105,7 +105,11 @@ template<typename Accessors>
 class base_splay_tree : protected Accessors
 {
 public:
+#if __GNUC__ > 4
   using typename Accessors::node_type;
+#else
+  using node_type = typename Accessors::node_type;
+#endif
 
   // INDEX is either 0 or 1.  If INDEX is 0, insert CHILD immediately
   // before NODE, otherwise insert CHILD immediately after NODE.
@@ -148,7 +152,11 @@ class rooted_splay_tree : public base_sp
   using parent = base_splay_tree<Accessors>;
 
 public:
+#if __GNUC__ > 4
   using typename Accessors::node_type;
+#else
+  using node_type = typename Accessors::node_type;
+#endif
 
 protected:
   // The root of the splay tree, or node_type () if the tree is empty.
@@ -409,7 +417,11 @@ class rootless_splay_tree
 public:
   using rooted = rooted_splay_tree<full_accessors>;
 
+#if __GNUC__ > 4
   using typename Accessors::node_type;
+#else
+  using node_type = typename Accessors::node_type;
+#endif
 
   // Remove NODE from the splay tree.  Return the node that replaces it,
   // or null if NODE had no children.
